libs/model/src/policies/utils/community-indexer-utils.ts

```diff
@@ -0,0 +1,268 @@
+    retryCondition: (error: AxiosError) =>
+      error.response?.status === 429 || (error.response?.status ?? 0) >= 500,
+  });
+
+  let pageNum = 1;
+
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const url = `https://www.clanker.world/api/tokens?sort=${desc ? 'desc' : 'asc'}&page=${pageNum}&pair=all&partner=all&presale=all`;
+    log.debug(`fetching: ${url}`);
+
+    try {
+      // Use the axios instance with retry logic
+      const res = await axiosInstance.get<{
+        data: Array<z.infer<typeof ClankerToken>>;
+        hasMore: boolean;
+      }>(url);
+
+      // If no tokens were returned, we're done
+      if (!res.data.hasMore || res.data.data.length === 0) {
+        log.debug('No more tokens found');
+        break;
+      }
+
+      // Filter tokens by cutoffDate
+      const validTokens = res.data.data.filter((t) =>
+        moment(t.created_at).isAfter(cutoffDate),
+      );
+
+      // Yield valid tokens
+      if (validTokens.length > 0) {
+        yield validTokens;
+      }
+
+      // Check if the oldest token is older than our cutoff date
+      const oldestToken = res.data.data[res.data.data.length - 1];
+      if (moment(oldestToken.created_at).isBefore(cutoffDate)) {
+        break;
+      }
+
+      // Move to the next page
+      pageNum++;
+    } catch (err: unknown) {
+      log.error(`Error fetching clanker tokens: ${(err as Error).message}`);
+      throw err;
+    }
+  }
+}
+
+function formatCommunityName(input: string) {
+  return (
+    input
+      // eslint-disable-next-line no-useless-escape
+      .replace(/[^a-zA-Z0-9!@#&():_$\/\\|.\- ]+/g, '') // Keep only allowed characters
+      .replace(/\s+/g, ' ') // Replace multiple spaces with one
+      .trim()
+  ); // Trim leading and trailing spaces
+}
+
+/**
+ * Generates a compliant community name and unique ID based on the provided community name.
+ *
+ * @param {string} name - The original community name.
+ * @returns {Promise<
+ *   | { id: string; name: string; error: null }
+ *   | { id: null; name: null; error: string }
+ * >}
+ */
+export async function generateUniqueId(
+  name: string,
+): Promise<
+  | { id: string; name: string; error: null }
+  | { id: null; name: null; error: string }
+> {
+  const communityName = formatCommunityName(name.trim());
+  if (communityName.length <= 3) {
+    return {
+      id: null,
+      name: null,
+      error: `formatted community name too short: original="${name}" formatted="${communityName}"`,
+    };
+  }
+  const baseId = lo.kebabCase(communityName);
+  let idCandidate = baseId;
+
+  const existingBase = await models.Community.findOne({
+    where: { id: idCandidate },
+  });
+  if (!existingBase) {
+    return {
+      id: idCandidate,
+      name: communityName,
+      error: null,
+    };
+  }
+
+  let counter = 2;
+  while (counter < 100) {
+    idCandidate = `${baseId}-${counter}`;
+    const existing = await models.Community.findOne({
```

Comment:
How often is this expected to return a community? Fine if rare, otherwise should pre-generate the ids and do a single query.